File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int l = 0;
        int r = 0;
        int s_index = -1;
        int hash[256] = {0};
        int count = 0;
        int minlen = INT_MAX;
        int n = s.size();
        int m = t.size();

        for(int i = 0; i < m; i++){
            hash[t[i]]++;
        }

        while(r < s.size()){
            if(hash[s[r]] > 0) count++;
            hash[s[r]]--;
            while(count == m){
                int lenn = r-l+1;
                if(minlen > lenn){
                    minlen = lenn;
                    s_index = l;
                }
                hash[s[l]]++;
                if(hash[s[l]] > 0) count--;
                l++;
            }
            r++;
        }

        if(s_index == -1) return "";
        return s.substr(s_index, minlen);
    }
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        l = 0
        r = 0
        min_len = 1e9
        hash_array = [0] * 256
        n = len(s)
        m = len(t)
        count = 0
        s_index = -1
        for i in range(m):
            hash_array[ord(t[i])] += 1

        while r < n:
            if hash_array[ord(s[r])] > 0:
                count += 1
            hash_array[ord(s[r])] -= 1

            while count == m:
                if (r-l+1) < min_len:
                    min_len = r-l+1
                    s_index = l
                hash_array[ord(s[l])] += 1
                if hash_array[ord(s[l])] > 0:
                    count -= 1
                l+=1

            r+=1
        if s_index == -1:
            return ""
        return s[s_index : s_index + min_len]